src/selectors/entities/channels.js

```diff
@@ -53,30 +54,39 @@ export {
     getMyCurrentChannelMembership,
 };
 
-export function getAllChannels(state) {
+import type {GlobalState} from '../../types/store';
+import type {Channel, ChannelStats, ChannelMembership} from '../../types/channels';
+import type {UsersState, UserProfile} from '../../types/users';
+import type {PreferenceType} from '../../types/preferences';
+import type {Post} from '../../types/posts';
+import type {TeamMembership} from '../../types/teams';
+
+type SortingType = 'recent' | 'alpha';
+
+export function getAllChannels(state: GlobalState): {[string]: Channel} {
     return state.entities.channels.channels;
 }
 
-export function getAllChannelStats(state) {
+export function getAllChannelStats(state: GlobalState): {[string]: ChannelStats} {
     return state.entities.channels.stats;
 }
 
-export function getChannelsInTeam(state) {
+export function getChannelsInTeam(state: GlobalState): {[string]: Array<string>} {
     return state.entities.channels.channelsInTeam;
 }
 
-export const getDirectChannelsSet = createSelector(
+export const getDirectChannelsSet: (GlobalState) => Set<string> = createSelector(
     getChannelsInTeam,
-    (channelsInTeam) => {
-        return channelsInTeam[''] || new Set();
+    (channelsInTeam: {[string]: Array<string>}): Set<string> => {
```

Comment:
Actually we can skip almost annotations for args and returns with reusable type utils.

For example:

```js
type Selector<T> = typeof createSelector<GlobalState, *, *, T>

export const getDirectChannelsSet: Selector<RelationOneToMany<Team, Channel>> = createSelector(
    getChannelsInTeam,
    (channelsInTeam) => {
        return new Set(channelsInTeam['']) || new Set();
    }
);
```
This code is perfectly type-safe

Reply:
Yes, I'm trying to be super explicit with the types. I love the type inference, but sometimes is more permissive than me.

Reply:
I'm OK to use annotation manually, 

This is what createSelector looks like:

```js
  declare type SelectorCreator = {
    <TState, TProps, TResult, T1>(
      selector1: InputSelector<TState, TProps, T1>,
      resultFunc: (arg1: T1) => TResult
    ): OutputSelector<TState, TProps, TResult>,
    <TState, TProps, TResult, T1>(
      selectors: [InputSelector<TState, TProps, T1>],
      resultFunc: (arg1: T1) => TResult
    ): OutputSelector<TState, TProps, TResult>,
```

since we use library and library, type annotation should depend on library type, even it is manuall annotation.

And some type annotations may need to be defined repeatedly.

1. TState is always GlobalState
2. Annotation for TResult (Set<string> in this case)

The suggested type utility `type Selector<T> = typeof createSelector<GlobalState, *, T, *>`.

[existential type](https://flow.org/en/docs/types/utilities/#toc-existential-type) is a most powerful feature (that TypeScript not have), use an existential type as a placeholder, it is inferred to be a later assigned type, not an any. so you still use mannual type annotation for TProps and T(N)

Reply:
Yes, this makes sense, I would suggest to include this in another PR, this is already complex enough. And I don't want to add extra complexity.